synapse/handlers/e2e_keys.py

```diff
@@ -636,7 +911,58 @@ def _check_cross_signing_key(key, user_id, key_type, signing_key=None):
             )
 
 
+def _check_device_signature(user_id, verify_key, signed_device, stored_device):
+    """Check that a device signature is correct and matches the copy of the device
+    that we have.  Throws an exception if an error is detected.
+
+    Args:
+        user_id (str): the user ID whose signature is being checked
+        verify_key (VerifyKey): the key to verify the device with
+        signed_device (dict): the signed device data
+        stored_device (dict): our previous copy of the device
+    """
+
+    key_id = "%s:%s" % (verify_key.alg, verify_key.version)
+
+    # make sure the device is signed
+    if (
+        "signatures" not in signed_device
+        or user_id not in signed_device["signatures"]
+        or key_id not in signed_device["signatures"][user_id]
+    ):
+        logger.error("upload signature: user not found in signatures")
+        raise SynapseError(400, "Invalid signature", Codes.INVALID_SIGNATURE)
+
+    signature = signed_device["signatures"][user_id][key_id]
+
+    # make sure that the device submitted matches what we have stored
+    del signed_device["signatures"]
+    # use pop to avoid exception if key doesn't exist
+    signed_device.pop("unsigned", None)
+    stored_device.pop("signatures", None)
+    stored_device.pop("unsigned", None)
+    if signed_device != stored_device:
```

Comment:
I'm not sure this does what you want it to do, though tbh I'm not sure what you want it to do. 

comparisons are shallow in python...

Reply:
It seems to be deep

```python
>>> x = {"a": {"b": {"c": "d"}}}
>>> x == {"a": {"b": {"c": "d"}}}
True
>>> x == {"a": {"b": {"c": "e"}}}
False
```

Reply:
you're right. I'm a crank.